File: libsrc/calc_covar.f90
module calc_covar
use, intrinsic :: iso_fortran_env, only : real64
use, intrinsic :: iso_c_binding, only : c_double, c_float, c_int

!use layers,       only: model_error, reading_error
!use const_raddeg

implicit none

double precision, parameter :: RAD2DEG=57.295779513082320876

contains

integer function eigv_sym4(a,w)
! Computes the eigenvalues and eigenvectors of a 4x4 symmetric matrix using LAPACK's dsyev.
! The matrix 'a' (dimension 4x4) is stored in column‐major order.
! On exit, if successful, a contains the orthonormal eigenvectors and w holds the eigenvalues.

integer, parameter :: n = 4
integer, parameter :: lda = n
real(real64), dimension(lda,n) :: a
real(real64), dimension(n)     :: w

integer, parameter :: lwmax = 200
integer            :: info, lwork
real(real64), dimension(lwmax) :: work

!  evaluate eigenvalues
!
![in]	JOBZ	CHARACTER*1
!          = 'N':  Compute eigenvalues only;
!          = 'V':  Compute eigenvalues and eigenvectors.
!
![in]	UPLO	CHARACTER*1
!          = 'U':  Upper triangle of A is stored;
!          = 'L':  Lower triangle of A is stored.
!
![in]	N	INTEGER
!          The order of the matrix A.  N >= 0.
!
![in,out]	A
!          A is DOUBLE PRECISION array, dimension (LDA, N)
!          On entry, the symmetric matrix A.  If UPLO = 'U', the
!          leading N-by-N upper triangular part of A contains the
!          upper triangular part of the matrix A.  If UPLO = 'L',
!          the leading N-by-N lower triangular part of A contains
!          the lower triangular part of the matrix A.
!          On exit, if JOBZ = 'V', then if INFO = 0, A contains the
!          orthonormal eigenvectors of the matrix A.
!          If JOBZ = 'N', then on exit the lower triangle (if UPLO='L')
!          or the upper triangle (if UPLO='U') of A, including the
!          diagonal, is destroyed.
!
![in]	LDA	INTEGER
!          The leading dimension of the array A.  LDA >= max(1,N).
!
![out]	W
!          W is DOUBLE PRECISION array, dimension (N)
!          If INFO = 0, the eigenvalues in ascending order.
!
![out]	WORK
!          WORK is DOUBLE PRECISION array, dimension (MAX(1,LWORK))
!          On exit, if INFO = 0, WORK(1) returns the optimal LWORK.
!
![in]	LWORK	INTEGER
!          The length of the array WORK.  LWORK >= max(1,3*N-1).
!          For optimal efficiency, LWORK >= (NB+2)*N,
!          where NB is the blocksize for DSYTRD returned by ILAENV.
!
!          If LWORK = -1, then a workspace query is assumed; the routine
!          only calculates the optimal size of the WORK array, returns
!          this value as the first entry of the WORK array, and no error
!          message related to LWORK is issued by XERBLA.
!
![out]	INFO	INTEGER
!          = 0:  successful exit
!          < 0:  if INFO = -i, the i-th argument had an illegal value
!          > 0:  if INFO = i, the algorithm failed to converge; i
!                off-diagonal elements of an intermediate tridiagonal
!                form did not converge to zero.

      lwork = -1
!     call dsyev( 'Vectors', 'Upper', n, a, lda, w, work, lwork, info )
      call dsyev( 'Vectors', 'Lower', n, a, lda, w, work, lwork, info )
      lwork = int( work( 1 ) )
!write(*,*) lwork
!      call dsyev( 'Vectors', 'Upper', n, a, lda, w, work, lwork, info )
      call dsyev( 'Vectors', 'Lower', n, a, lda, w, work, lwork, info )
      if( info.gt.0 ) then
         write(*,*)'The algorithm failed to compute eigenvalues.'
!         stop
      end if
      eigv_sym4=info

end function eigv_sym4

subroutine cov_matrix2(cov,m,n,resw,Gw,w,reading_error,model_error) bind(C,name='cov_matrix2')
!covariance matrix
real(c_double), dimension(n,n), intent(OUT) :: cov
integer(c_int), intent(IN), value          :: m,n
real(c_double), dimension(m), intent(IN)   :: resw
real(c_double), dimension(m), intent(IN)   :: w
real(c_double), dimension(m,n), intent(IN) :: Gw
real(c_double), intent(IN), value :: reading_error,model_error

real(real64), parameter :: kappa=1.0d-08

real(real64), dimension(n,n) :: GG,eigV,covM,pmat !,eigL
real(real64), dimension(n) :: eig_l
real(real64)               :: varD
integer                    :: marrv
real(real64)               :: sumw
integer :: j
integer :: info

!GG=G*G'
GG=MATMUL(TRANSPOSE(Gw),Gw)

!eigenvalues (ortogonal) decomposition
!V*diag(l)*V'=GG
eigV=GG
info = eigv_sym4(eigV,eig_l)
!test
!   eigL=0.0
!   do i=1,4
!         eigL(i,i)=eig_l(i)
!   end do
!   GG=MATMUL(eigV,MATMUL(eigL,TRANSPOSE(eigV)))

if (minval(eig_l)/maxval(eig_l)<kappa) then
        eig_l=eig_l+kappa
end if

!covariance form eigenvalues
!C=V*diag(1/l)*V'
do j=1,n
    pmat(j,:)=eigV(:,j)/eig_l(j)
end do
covM=MATMUL(eigV,pmat)
!test
!   eigL=0.0
!   do i=1,4
!         eigL(i,i)=1/eig_l(i)
!   end do
!   covM=MATMUL((eigV),MATMUL(eigL,TRANSPOSE(eigV)))


sumw=sum(w)
marrv=count(w>0)
varD=dot_product(resw,resw)/sumw
if (marrv>4) then
        varD=varD*marrv/(marrv-4)
else
        varD=varD*marrv
end if

! reading error
   if (varD .lt. reading_error**2) then
      varD = reading_error**2
   endif

! model error
   varD=varD+model_error**2

cov=covM*varD
end subroutine cov_matrix2

end module calc_covar

File: libsrc/calc_covar.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <numeric>
#include <lapacke.h>

// Define constants from layers and const_raddeg modules
const double model_error = 0.2;    // value assumed from layers module
const double reading_error = 0.1;  // value assumed from layers module
const double RAD2DEG = 180.0 / M_PI; // constant from const_raddeg module

// Typedef for a matrix stored in column‐major order.
// A matrix is represented by a 1D vector of doubles. For a matrix with nr rows and nc columns,
// the element (i, j) (with 0-based indexing) is at index: i + j * nr.
typedef std::vector<double> Matrix;

// Inline helper functions for accessing and setting matrix elements in column‐major order.
inline double get(const Matrix &A, int nr, int i, int j) {
    return A[i + j * nr];
}
inline void set(Matrix &A, int nr, int i, int j, double value) {
    A[i + j * nr] = value;
}

// Helper function: Multiply two square matrices A and B (dimension n x n) stored in column‐major order.
Matrix matmul_square(const Matrix &A, const Matrix &B, int n) {
    Matrix C(n * n, 0.0);
    for (int j = 0; j < n; j++) {
        for (int i = 0; i < n; i++) {
            double sum = 0.0;
            for (int k = 0; k < n; k++) {
                sum += get(A, n, i, k) * get(B, n, k, j);
            }
            set(C, n, i, j, sum);
        }
    }
    return C;
}

// Helper function: Compute matrix product GG = MATMUL(TRANSPOSE(Gw), Gw)
// Gw is an m x n matrix stored in column‐major order. The result GG is n x n.
Matrix matmul_transpose(const Matrix &Gw, int m, int n) {
    Matrix GG(n * n, 0.0);
    for (int j = 0; j < n; j++) {
        for (int i = 0; i < n; i++) {
            double sum = 0.0;
            for (int k = 0; k < m; k++) {
                // In Gw, element (k, i) is at index k + i * m, (k, j) is at k + j * m.
                sum += Gw[k + i * m] * Gw[k + j * m];
            }
            set(GG, n, i, j, sum);
        }
    }
    return GG;
}

namespace calc_covar {

    // Subroutine eigv_sym4
    // Computes the eigenvalues and eigenvectors of a 4x4 symmetric matrix using LAPACK's dsyev.
    // The matrix 'a' (dimension 4x4) is stored in column‐major order.
    // On exit, if successful, a contains the orthonormal eigenvectors and w holds the eigenvalues.
    void eigv_sym4(Matrix &a, double w[4])
    {
        const int n = 4;
        const int lda = n;
        int info;
        int lwork = -1;
        double work_query;
        // Workspace query for optimal lwork
        info = LAPACKE_dsyev_work(LAPACK_COL_MAJOR, 'V', 'L', n, a.data(), lda, w, &work_query, lwork);
        lwork = static_cast<int>(work_query);
        std::vector<double> work(lwork);
        // Evaluate eigenvalues and eigenvectors
        info = LAPACKE_dsyev_work(LAPACK_COL_MAJOR, 'V', 'L', n, a.data(), lda, w, work.data(), lwork);
        if (info > 0) {
            std::cout << "The algorithm failed to compute eigenvalues." << std::endl;
            std::exit(EXIT_FAILURE);
        }
    }

    // Function cov_matrix1
    // Computes the covariance matrix.
    // Parameters:
    //   m     : INTEGER, number of rows for resw, w, and Gw.
    //   n     : INTEGER, number of columns for Gw and dimension of the covariance matrix (n x n).
    //   resw  : REAL(real64) array of dimension m.
    //   Gw    : REAL(real64) matrix of dimension (m, n) stored in column‐major order.
    //   w     : REAL(real64) array of dimension m.
    // Returns:
    //   cov_matrix1 : REAL(real64) matrix of dimension (n, n) stored in column‐major order.
    Matrix cov_matrix1(int m, int n, const std::vector<double> &resw, const std::vector<double> &w, const Matrix &Gw)
    {
        //--- Begin function cov_matrix1 ---
        // For this implementation, n must equal 4 as eigv_sym4 operates on a 4x4 matrix.
        if (n != 4) {
            std::cerr << "cov_matrix1 error: n must be 4." << std::endl;
            std::exit(EXIT_FAILURE);
        }
        const double kappa = 1.0e-8;
        // Compute GG = MATMUL(TRANSPOSE(Gw), Gw)
        Matrix GG = matmul_transpose(Gw, m, n); // GG is n x n.
        Matrix eigV = GG;  // Copy of GG to be used for eigen-decomposition.
        double eig_l[4];
        // Compute eigenvalues and eigenvectors of eigV (which is GG)
        eigv_sym4(eigV, eig_l);
        // Adjust eigenvalues if their ratio is below kappa.
        double min_eig = eig_l[0], max_eig = eig_l[0];
        for (int i = 1; i < 4; i++) {
            if (eig_l[i] < min_eig)
                min_eig = eig_l[i];
            if (eig_l[i] > max_eig)
                max_eig = eig_l[i];
        }
        if (min_eig / max_eig < kappa) {
            for (int i = 0; i < 4; i++) {
                eig_l[i] = eig_l[i] + kappa;
            }
        }
        // Build pmat such that pmat(j,:) = eigV(:,j)/eig_l(j)
        // Here, we set the j-th row of pmat from the j-th column of eigV.
        Matrix pmat(4 * 4, 0.0);
        for (int j = 0; j < 4; j++) {
            for (int i = 0; i < 4; i++) {
                // In eigV, element (i,j) is accessed as get(eigV, 4, i, j)
                // We assign this to pmat at position (j,i).
                double val = get(eigV, 4, i, j) / eig_l[j];
                set(pmat, 4, j, i, val);
            }
        }
        // Compute covariance matrix in eigenvalue form: covM = MATMUL(eigV, pmat)
        Matrix covM = matmul_square(eigV, pmat, 4);
        // Compute sumw = sum of elements of w
        double sumw = std::accumulate(w.begin(), w.end(), 0.0);
        // Count number of positive elements in w
        int marrv = std::count_if(w.begin(), w.end(), [](double val) { return val > 0.0; });
        // Compute variance factor varD from resw and w.
        double dot = 0.0;
        for (int i = 0; i < m; i++) {
            dot += resw[i] * resw[i];
        }
        double varD = dot / sumw;
        if (marrv > 4) {
            varD = varD * marrv / (marrv - 4);
        } else {
            varD = varD * marrv;
        }
        // reading error: ensure varD is not below reading_error².
        if (varD < reading_error * reading_error) {
            varD = reading_error * reading_error;
        }
        // model error is added to the variance.
        varD = varD + model_error * model_error;
        // Multiply the covariance matrix by the variance factor.
        for (auto &val : covM) {
            val *= varD;
        }
        return covM;
        //--- End function cov_matrix1 ---
    }

    // Subroutine get_errellipse
    // Computes the error ellipse parameters from a 4x4 covariance matrix 'co' stored in column‐major order.
    // Output parameters:
    //   dxer, dyer : The square roots of the absolute values of co(1,1) and co(2,2).
    //   dzer, dter : The square roots of the absolute values of co(3,3) and co(4,4).
    //   l1, l2     : The lengths of the major and minor axes of the error ellipse.
    //   theta      : The orientation angle of the error ellipse (in degrees).
    void get_errellipse(const Matrix &co, double &dxer, double &dyer, double &dzer, double &dter,
                        double &l1, double &l2, double &theta)
    {
        //--- Begin subroutine get_errellipse ---
        // Compute the determinant from the 2x2 upper-left block of co.
        double deter = get(co, 4, 0, 0) * get(co, 4, 1, 1) - get(co, 4, 0, 1) * get(co, 4, 1, 0);
        double d11 = get(co, 4, 1, 1) / deter;
        double d22 = get(co, 4, 0, 0) / deter;
        double d21 = -get(co, 4, 1, 0) / deter;
        theta = std::atan(2 * d21 / (d11 - d22)) / 2.0;
        double cos_theta = std::cos(theta);
        double sin_theta = std::sin(theta);
        double al = d11 * cos_theta * cos_theta + 2 * d21 * cos_theta * sin_theta + d22 * sin_theta * sin_theta;
        double bl = d11 * sin_theta * sin_theta - 2 * d21 * cos_theta * sin_theta + d22 * cos_theta * cos_theta;
        l1 = std::sqrt(1.0 / al);
        l2 = std::sqrt(1.0 / bl);
        // Convert theta to degrees.
        theta = theta * RAD2DEG;
        // Adjust if l2 is greater than l1.
        if (l2 > l1) {
            double tl = l1;
            l1 = l2;
            l2 = tl;
            theta = theta + 90.0;
        }
        dxer = std::sqrt(std::fabs(get(co, 4, 0, 0)));
        dyer = std::sqrt(std::fabs(get(co, 4, 1, 1)));
        dzer = std::sqrt(std::fabs(get(co, 4, 2, 2)));
        dter = std::sqrt(std::fabs(get(co, 4, 3, 3)));
        //--- End subroutine get_errellipse ---
    }

} // end namespace calc_covar

// Example main function demonstrating usage (can be removed or modified as needed)
int main() {
    // Example usage of eigv_sym4:
    // Define a 4x4 symmetric matrix (column-major order)
    Matrix A = {
        4.0, 1.0, 2.0, 0.0,  // first column
        1.0, 3.0, 0.0, 1.0,  // second column
        2.0, 0.0, 2.0, 1.0,  // third column
        0.0, 1.0, 1.0, 1.0   // fourth column
    };
    double w[4];
    calc_covar::eigv_sym4(A, w);
    std::cout << "Eigenvalues:" << std::endl;
    for (int i = 0; i < 4; i++) {
        std::cout << w[i] << std::endl;
    }

    // Example usage of cov_matrix1:
    // Assume m = 5 and n = 4.
    int m = 5, n = 4;
    // Example resw and w vectors.
    std::vector<double> resw = {0.5, 0.6, 0.7, 0.8, 0.9};
    std::vector<double> warr = {1.0, 2.0, 3.0, 4.0, 5.0};
    // Example Gw: an m x n matrix stored in column-major order.
    // For simplicity we initialize Gw with 5 rows and 4 columns.
    Matrix Gw = {
        // Column 1:
        1.0, 2.0, 3.0, 4.0, 5.0,
        // Column 2:
        2.0, 3.0, 4.0, 5.0, 6.0,
        // Column 3:
        3.0, 4.0, 5.0, 6.0, 7.0,
        // Column 4:
        4.0, 5.0, 6.0, 7.0, 8.0
    };
    Matrix covMat = calc_covar::cov_matrix1(m, n, resw, warr, Gw);
    std::cout << "Covariance Matrix (4x4):" << std::endl;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            std::cout << get(covMat, n, i, j) << " ";
        }
        std::cout << std::endl;
    }

    // Example usage of get_errellipse:
    // Use covMat as the 4x4 covariance matrix "co".
    double dxer, dyer, dzer, dter, l1, l2, theta;
    calc_covar::get_errellipse(covMat, dxer, dyer, dzer, dter, l1, l2, theta);
    std::cout << "Error Ellipse Parameters:" << std::endl;
    std::cout << "dxer = " << dxer << ", dyer = " << dyer << std::endl;
    std::cout << "dzer = " << dzer << ", dter = " << dter << std::endl;
    std::cout << "l1 (major axis) = " << l1 << ", l2 (minor axis) = " << l2 << std::endl;
    std::cout << "theta = " << theta << " degrees" << std::endl;

    return 0;
}

